content/articles/hashing-in-data-structure/index.md

```diff
@@ -0,0 +1,314 @@
+        cs++;
+
+        if(cs/(1.0*ts)  > 0.6){
+            reHash();
+        }
+
+    }
+
+    node* search(string key){
+        int i=hashfn(key);
+        node*head= arr[i];
+        while(head){
+            if(head->name==key){
+               return head;
+                break;
+            }
+            head=head->next;
+        }
+        if(head==NULL){
+            cout<<"not exist";
+        }
+        return NULL;
+    }
+
+    void print(){
+        for(int i=0;i<ts;i++){
+            node* head= arr[i];
+            while(head){
+                cout<<head->name<<"-->"<<head->value;
+                head=head->next;
+                 cout<<endl;
+            }
+
+        }
+    }
+
+    int& operator[](string key){
+        node* ans=search(key);
+        if(ans){
+            return ans->value;
+        }
+        else{
+            int agrbagevalue;
+            insert(key,agrbagevalue);
+            ans= search(key);
+            return ans->value;
+        }
+
+    }
+
+    void Delete(string key){
+        int i=hashfn(key);
+        node* head= arr[i];
+        node* trail=NULL;
+
+            while(head){
+
+                    //first
+                    if(head->name==key && trail==NULL){
+                        arr[i]=NULL;
+                        delete head;
+                    }
+                    //in end
+                    if(head->name==key && trail!=NULL && head->next==NULL){
+                        trail->next=NULL;
+                        delete head;
+                    }
+                    //mid
+                    if(head->name==key && head->next!=NULL){
+                        node*ptr =head;
+                        head=head->next;
+                        delete ptr;
+                        arr[i]=head;
+
+                    }
+                    trail =head;
+                    head=head->next;
+            }
+        return;
+    }
+
+};
+
+int main(){
+    hashmap h;
+    h.insert("Alphonso_Mango",100);
+    h.insert("Kiwi",150);
+    h.insert("Banana",200);
+    h.insert("WaterMelon",180);
+    h.print();
+
+    node* x=  h.search("Kiwi");
+    cout<<"SEARCH RESULT"<<endl;
+    cout<<x->name<<" "<<x->value;
+
+}
+
+```
+
+The complete execution of the above code can be found [here](https://replit.com/@Anubhavb11/HashMap#main.cpp).
+=======
+
+#### Output
+
+![output](./Capture.PNG)
+
+* In the output you can see we have inserted different key-value pairs into the custom hash table build by us. Try to experiment with these values and you will find collisions also and rehashing too if there are many collisions.
+
+* Searching happening in O(1) because of the hash table.
+
+#### Conclusion
+
+With the help of hash tables, we can insert, search and delete in O(1) time which is a great achievement. Hash tables are widely used for making our code more efficient and faster. Here are some problems of data structures you can try and think the approach using hash tables.
```

Comment:
Searching happens in 0(1) and 0(1) time? What do you mean by this?

Reply:
@Bayler  I have tried to make it more clear you can have a look.